### DISCUSSION TITLE
Why there is no a built-in async lock in .NET?

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/runtime#83019] [Q&A]
I know there's a 3rd library called [`AsyncEx`](https://github.com/StephenCleary/AsyncEx). But why still there's no a built-in one even in .NET 7?

**Comment by En3Tho** [comment]
There are. One of the examples: https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-7.0

**Reply by GF-Huang** [reply]
This is only an approximate substitute. In other words, you must use it with `try/finally` pattern, it is not as convenient to use as AsyncLock of AsyncEx. I means if it would be nice to have built-in syntactic sugar like:
```cs
await lock(xxx) {
    ...
    await objA.XxxAsync();
    ...
    await objB.XxxAsync();
    ...
}
```

**Reply by Xyncgas** [reply]
Alternatively you don't have to do anything because your code is already async if you wait synchronously the manager is still going to switch context and execute some other task and come back later, that means as long you are using async you can use synchronous lock in system that has only one thread like Blazor.Wasm and it still works

**Reply by GF-Huang** [reply]
But you can't `await` inside `lock` statement.

**Reply by Xyncgas** [reply]
> But you can't `await` inside `lock` statement.

ohh
```Interlocked.CompareExchange``` Use this for atomic operations then, it locks on reference type like the lock statement but it's just one line that will change the value with atomic operation as long the value is a reference type variable, it's faster than lock because it doesn't use the object header to lock stuff it just locks

**Reply by Clockwork-Muse** [reply]
> , it locks on reference type like the lock statement but it's just one line that will change the value with atomic operation as long the value is a reference type variable, 

1. It performs no locking.  That's kind of the point of entire `Interlocked` class.
2. The methods in `Interlocked` do not allow for atomic operations over multiple values.  If you need to operate on multiple values you need to use a `lock` (or be really good at lock-free programming and be using algorithms and data structures that support it, which isn't a given).

**Reply by Clockwork-Muse** [reply]
> But you can't `await` inside `lock` statement.

While this is true, you might be able to restructure your code to move the `await` outside of the lock.

That said, mixing `async` and locks can be dangerous, since the two have fundamentally different approaches to handling resource conflict.

**Reply by Xyncgas** [reply]
>     1. It performs no locking.  That's kind of the point of entire `Interlocked` class.

Alright
It's better this way because you don't need to lock an object using it's header, which value types doesn't have unless you box it, while you can perform atomic operation on the pointer which is very fast
Int, byte, their operations are atomic in .NET unlike c / c++ already by the way

>     2. The methods in `Interlocked` do not allow for atomic operations over multiple values. 

just write this code multiple times then or abstract this line of code so it can work with multiple values

**Reply by GF-Huang** [reply]
> just write this code multiple times then or abstract this line of code so it can work with multiple values

Oh, my god, You can't guarantee transactionality by writing this way.

**Reply by Xyncgas** [reply]
> Oh, my god, You can't guarantee transactionality by writing this way.

what's wrong, just write 
```obj.ReferenceEquals(comparand, Interlocked.CompareExchange(&value, newValue, comparand))``` in a loop for example for a single object or multiple lines like this in a loop for multiple objects you want to lock

in such loop, you can yield or await like you normally would, so you are not blocking, but you are. 

For example you have multiple objects you want to operate on and make sure no body is racing with you, well you already have this promised from ```Interlocked.CompareExchange``` and it happens very fast because it works with the pointer, if you simply wanted a copy that no body can touch you can use ```Volatile.Read &value``` but if you are writing / updating an object this is completely fine.

The point of having a loop is because ```obj.ReferenceEquals(comparand, Interlocked.CompareExchange(&value, newValue, comparand))``` happens so fast but it returns false if you are indeed being raced, in that case you retry in a loop and that's what you would call blocking but this is faster than semaphorelite because it's accelerated by hardware and it 100% works, where in this loop you can yield your async code

**Reply by GF-Huang** [reply]
Seems you don't understand what is transaction/transactional. How do you do the following by your `Interlocked.CompareExchange` approach?

```cs
using AsyncEx;

...

private readonly AsyncLock _lock = new AsyncLock();

...

using (await _lock.LockAsync()) {
    ...
    await objA.XxxAsync();
    ...
    await objB.XxxAsync();
    ...
}
```

Or let me put it more bluntly:

```cs
private readonly Dictionary<string, object> _dictA; // not thread-safety dictionary
private readonly Dictionary<string, object> _dictB; // not thread-safety dictionary
private readonly AsyncLock _lock = new AsyncLock();

...

public Task UpdateAsync(string parameter) {
    using (await _lock.LockAsync()) {
        ...
        var resultA = await objA.XxxAsync();
        _dictA[parameter] = resultA;
        ...
        var resultB = await objB.XxxAsync();
        _dictB[parameter] = resultB;
        ...
    }
}
```

**Reply by Clockwork-Muse** [reply]
@GF-Huang - 
In theory, you could do something like this:
```csharp
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Threading;
					
public class Program
{
	private object _lock = new object();
	
	private object? _a = null;
	private object? _b = null;

	public void Update(string parameter, object a, object b) {
		object savedLock = null;
		// Spin until object acquired
		while((savedLock = Interlocked.Exchange(ref _lock, null)) == null);
		
		_a = a;
		_b = b;
		
		Interlocked.Exchange(ref _lock, savedLock);
	}
}
```
... which would be safe, due to memory model guarantees.  It should be applicable to setting elements in dictionaries as well.

That said, unless you can guarantee the lock is held for a _very_ short period of time, this isn't necessarily beneficial, as `lock` is already fairly well optimized.  Note too that this is a spin lock, which means the CPU is actively being used, which can prevent other things from executing.

@Xyncgas - I have no idea if the memory model guarantees apply over async suspension points, especially because in many cases the resumed execution will be on another thread.  But the larger problem is that for any serious `async` work (anything calling I/O, especially over a network), the lock will be held for a **long** period of time, which means that the thread is being burned (and may even mean that the program live locks, if the thread pool that should be servicing the suspended tasks decides not to allocate a new thread).

**Reply by GF-Huang** [reply]
@Clockwork-Muse Spin-lock is not good for network IO.

**Reply by Xyncgas** [reply]
> Spin-lock is not good for network IO.

is it because it's long running

**Reply by Xyncgas** [reply]
I suggest we can add await inside the while loop, so the while loop itself is checking whether it can continue...
Although ideal asynchronous should be resuming code on call back, by taking out the execution routine and pasting it to the event of readiness, instead of actively checking something is ready an approach to get activated by event is better

**Reply by Clockwork-Muse** [reply]
> is it because it's long running

It can be, yes (and in the case of the official `SpinLock(...)` method, it can be upgraded to an actual blocking lock, too).

> I suggest we can add await inside the while loop, so the while loop itself is checking whether it can continue

Not really.
The better thing is probably to change how a wider context is coded, because _generally_ you shouldn't mix the two paradigms.
If that's not really practical, using one of the (currently) third-party async-lock libraries is probably a better bet.

**Comment by Clockwork-Muse** [comment]
>  if you wait synchronously the manager is still going to switch context and execute some other task and come back later

`Sleep()`/`Yield()` yes, `lock` no.  Locks in C# are blocking (in contrast to Java, where it is a suspension point).

**Reply by Xyncgas** [reply]
If you are using a lock on an object, despite the common misconception that there is no indication of there being a point where your code would stop executing, .NET can still stop executing your code and switch the context to execute. It can because the codes are being executed in a virtual machine, just like java

Therefore being inside a lock statement doesn't ensure your code will execute without interruption

**Reply by Clockwork-Muse** [reply]
Sure, but that's not what I'm referring to.  
When you call `lock`, but the object you're referencing is already being used as a lock by another thread, the current thread blocks (Stops executing) until the lock is released.  This means that another thread (As in, OS thread) would need to be created first, which can take a while.

In contrast, Java (and some other languages) use virtual threads (... somewhat close to `Task` in concept, although Java doesn't have an `async` equivalent), so locks and waits are suspension points instead.

**Comment by AartBluestoke** [comment]
is there anything language wise stopping the runtime from implementing a block-on-lock as an async-like await, using the thread for something else for a short while? 
(i know it would be complicated to know what out of stack/cpu state would need to be persisted, but again, implementation detail, not a language one)

**Reply by Clockwork-Muse** [reply]
See [this question about green threads](https://github.com/dotnet/runtime/discussions/81462).

**Reply by Xyncgas** [reply]
Member Haskell
